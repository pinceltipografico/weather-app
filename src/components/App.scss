@import '~@/scss/mixins';

.weather-view {
  background-color: color(secondary);
  flex: 1 1 auto;
}
