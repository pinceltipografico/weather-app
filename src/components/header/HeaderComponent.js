import React, { Component } from 'react';
import logo from '../../assets/logo.svg';
import styles from './HeaderComponent.scss';
import PropTypes from 'prop-types';
import { Selects } from '../select/Selects';
import { citiesList } from '../../constants/citieslist';
import { WeatherConsumer } from '../../contexts/weatherContext';


class HeaderComponent extends Component {

  static propTypes = {
    title: PropTypes.string.isRequired
  }

  render () {
    return (
    <header className={styles.header}>
      <img src={logo} alt="logo" className={styles.logo} />  
      <h2 className={styles.title}>{this.props.title}</h2>
      
      <WeatherConsumer>
        { ({ fetchWeathers }) => (
          <Selects onChange={(value) => fetchWeathers(value)}
            label="Select a city"
            list={citiesList}/>
        )}
      </WeatherConsumer>
    </header>
    )
  }
}

export {
  HeaderComponent
}