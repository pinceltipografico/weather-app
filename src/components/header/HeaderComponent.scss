@import '~@/scss/mixins';

$white: #fff;

.header {
  @include flex-container;

  background-color: color(primary);
  flex: 1 1 auto;
  height: rem(100px);
  padding: 0 rem(20px);
  position: fixed;
  z-index: 2;
}

.logo {
  max-height: rem(80px);
}

.title {
  border-left: 1px dashed $white;
  border-right: 1px dashed $white;
  color: $white;
  margin: rem(10px) 0 0 rem(20px);
  padding: 0 rem(20px);
  
  @include breakpoint(phone-wide, max) {
    display: none;
  }
}
