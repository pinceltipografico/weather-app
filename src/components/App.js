import React, { Component } from 'react';
import { HeaderComponent } from './header/HeaderComponent';
import PropTypes from 'prop-types';
import styles from './App.scss';
import { BannerComponent } from './banner/BannerComponent';
import { WeatherListComponent } from './weatherlist/WeatherListComponent';
import { WeatherProvider } from '../contexts/weatherContext';
import { weatherApi } from '../constants/weatherapi';
import dayjs from 'dayjs';


class App extends Component {

  static propTypes = {
    apiKey: PropTypes.string.isRequired
  };

  state = {
    weatherList: [],
    currentCity: 'Vancouver',
    todayWeather: null,
    fetching: true
  };

  async componentDidMount () {
    await this.fetchWeathers('Vancouver');
  }

  /**
   * GROUP LIST BY DAY
   * TO SHOW MORE CONCISE ITEMS
   */
  formmatResults = (results) => {
    if(!results) {
      return
    }
    const {list} = results;
    const result = []
    const weatherList = list.reduce((previews, next) => {
      if(!previews.length) {
        return result
      }
      const day = dayjs(next.dt_txt).$D
      const group = previews.filter(i => dayjs(i.dt_txt).$D === day)
      if(group.length) {
        result.push(group)
      }
      return previews.filter(i => dayjs(i.dt_txt).$D !== day)
    }, list );

    this.setState({
      weatherList,
      todayWeather: weatherList[0][0],
      fetching: false
    })
  }

  /**
   * FETCH FROM API
   */
  fetchWeathers = async (city = 'London') => {
    const {endpoint, key} = weatherApi;
    const toGet = `${endpoint}?q=${city}&appid=${key}&units=imperial`;

    try {
      const response = await fetch(toGet,{
        method: 'GET'
      });
      const jsonResult = await response.json()
      return this.formmatResults(jsonResult);
    } catch (error) {
      this.setState({error});
    }
  }

  render () {

    const provider = {
      ...this.state,
      fetchWeathers: this.fetchWeathers.bind(this)
    };

    const {weatherList, fetching} = this.state

    return (
      <WeatherProvider value={provider}>

        <HeaderComponent title="Weather App" />

        {fetching &&
        <div style={{color: "#28a9b6", margin: '120px auto'}}>
            loading forecasts...
        </div>
        }

        {(!fetching && weatherList.length) &&
        <React.Fragment>

          <BannerComponent city={this.state.currentCity}
            todayWeather={this.state.todayWeather} />

          <section className={styles['weather-view']}>
            <WeatherListComponent list={this.state.weatherList}/>
          </section>
          
        </React.Fragment>
        }

      </WeatherProvider>
    )
  }
}

export default App
