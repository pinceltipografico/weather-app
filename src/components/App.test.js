import React from 'react';
import App from './App';
import { HeaderComponent } from './header/HeaderComponent';
import { shallow, mount } from 'enzyme';

describe('AppComponent', () => {

  let mountedApp = null
  const AppComponent = () => {
    if(!mountedApp){
      mountedApp = mount(
        <App apiKey="1c8fb84121cb361184855d25de956e74" />
      )
    }
    return mountedApp
  }

  it('should render app without error', () => {
    const component = shallow(<App apiKey="asdasdasd" />);
    expect(component).toMatchSnapshot()
  })

  it('should contains HeaderComponent', () => {
    expect(AppComponent().find(HeaderComponent).length).toBe(1)
  })

})
