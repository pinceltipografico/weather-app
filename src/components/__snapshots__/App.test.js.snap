// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should render app without error 1`] = `
<ContextProvider
  value={
    Object {
      "currentCity": "Vancouver",
      "fetchWeathers": [Function],
      "fetching": true,
      "todayWeather": null,
      "weatherList": Array [],
    }
  }
>
  <HeaderComponent
    title="Weather App"
  />
</ContextProvider>
`;
